### ISSUE TITLE
Unexpected Description Logics Generated by EvoLearner in OntoLearn

### ISSUE THREAD

**Issue by AjayUmakanth** [dice-group/Ontolearn#502]
I'm encountering an issue with the EvoLearner model in OntoLearn, where the generated Description Logics after fitting the model do not match the expected result.

Issue:
The model is supposed to generate a Description Logic expression that describes the Graph class having more than 2 has_structure relationships with instances of SuperStructure. However, the generated expressions do not reflect this.

Instead, the output is:

```
∀ has_structure.⊥ 0.30769
∀ has_structure.wheel 0.30769
¬SuperStructure 0.30769
```
Above is the Description logic and the corresponding score. I was expecting a logic like:

```
Graph ⊓ ∃has_structure.Thing > 2
```

This would define Graph as having more than two has_structure relationships with SuperStructure instances.
I am getting similar issues for different cardinality restrictions too, not just greater than 2.

Code Snippet and OWL Graph:
I have attached a simple code snippet and a small mock OWL graph to demonstrate the issue. Please let me know what could be the issue and if you need additional information

Code Snippet:

```
import os
from ontolearn.learning_problem import PosNegLPStandard
from ontolearn.owlapy.model import OWLNamedIndividual, IRI, OWLObjectProperty
from ontolearn.knowledge_base import KnowledgeBase
from ontolearn.concept_learner import EvoLearner
from ontolearn.metrics import F1


def load_graph_and_explain(owl_file_path: str, namespace: str = "http://example.org/"):
    """
    Load the OWL graph, set positive and negative examples, and feed them to EvoLearner.
    Args:
        owl_file_path (str): Path to the OWL file.
        namespace (str): Namespace for the individuals in the graph.
    """
    if not os.path.isfile(owl_file_path):
        raise FileNotFoundError(f"OWL file '{owl_file_path}' does not exist.")

    # Load the OWL graph as a KnowledgeBase
    kb = KnowledgeBase(path=owl_file_path)

    # Define the 'has_structure' object property
    has_structure_property = OWLObjectProperty(IRI.create(f"{namespace}has_structure"))

    positive_examples = []
    negative_examples = []

    # Retrieve all individuals in the knowledge base
    individuals = list(kb.individuals())

    # Classify individuals based on the number of 'has_structure' relationships
    for individual in individuals:
        # Get the number of `has_structure` values
        structure_count = sum(
            1 for _ in kb.get_object_property_values(individual, has_structure_property)
        )
        print(f"{individual} has {structure_count}")

        if structure_count > 2:
            positive_examples.append(individual)
        else:
            negative_examples.append(individual)

    if not positive_examples:
        raise ValueError("No positive examples found in the loaded graph.")

    # Convert to OWLNamedIndividual sets
    typed_pos = set(positive_examples)
    typed_neg = set(negative_examples)

    # Create the learning problem
    lp = PosNegLPStandard(pos=typed_pos, neg=typed_neg)

    # Initialize the EvoLearner
    model = EvoLearner(
        knowledge_base=kb,
        use_data_properties=True,
        max_runtime=60,
        num_generations=400,
        quality_func=F1(),
    )

    # Fit the model and retrieve hypotheses
    model.fit(lp)
    hypotheses = model.best_hypotheses(n=5)

    from ontolearn.owlapy.render import DLSyntaxObjectRenderer

    renderer = DLSyntaxObjectRenderer()
    # Print the description logic for each hypothesis
    print("Description Logics:")
    for hypothesis in hypotheses:
        print(renderer.render(hypothesis.concept), hypothesis.quality)  # Print the OWL class expression

if __name__ == "__main__":
    # Define the path to the OWL file generated earlier
    owl_file = "/Users/ajay/Documents/owlGraphs/custom.owl"  # Update with the correct path

    # Run the explanation process
    load_graph_and_explain(owl_file_path=owl_file)

```

OWL Graph Example:
```
<?xml version="1.0"?>
<rdf:RDF xmlns="http://www.w3.org/2002/07/owl#"
     xml:base="http://www.w3.org/2002/07/owl"
     xmlns:ns1="http://example.org/"
     xmlns:owl="http://www.w3.org/2002/07/owl#"
     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
     xmlns:xml="http://www.w3.org/XML/1998/namespace"
     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#">
    <Ontology/>
    
    <!-- Object Properties -->
    <ObjectProperty rdf:about="http://example.org/has_structure">
        <rdfs:domain rdf:resource="http://example.org/Graph"/>
        <rdfs:range rdf:resource="http://example.org/SuperStructure"/>
    </ObjectProperty>

    <!-- Classes -->
    <Class rdf:about="http://example.org/Graph"/>
    <Class rdf:about="http://example.org/SuperStructure"/>
    <Class rdf:about="http://example.org/grid">
        <rdfs:subClassOf rdf:resource="http://example.org/SuperStructure"/>
    </Class>
    <Class rdf:about="http://example.org/house">
        <rdfs:subClassOf rdf:resource="http://example.org/SuperStructure"/>
    </Class>
    <Class rdf:about="http://example.org/wheel">
        <rdfs:subClassOf rdf:resource="http://example.org/SuperStructure"/>
    </Class>

    <!-- Individuals with 0 structures -->
    <NamedIndividual rdf:about="http://example.org/graph_001">
        <rdf:type rdf:resource="http://example.org/Graph"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_002">
        <rdf:type rdf:resource="http://example.org/Graph"/>
    </NamedIndividual>

    <!-- Individuals with 1 structures -->
    <NamedIndividual rdf:about="http://example.org/graph_101">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/grid"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_102">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_103">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/wheel"/>
    </NamedIndividual>

    <!-- Individuals -->
    <!-- Individuals with exactly two structures -->
    <NamedIndividual rdf:about="http://example.org/graph_200">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/grid"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_201">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/wheel"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_202">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/grid"/>
        <ns1:has_structure rdf:resource="http://example.org/wheel"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_203">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
        <ns1:has_structure rdf:resource="http://example.org/wheel"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_204">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/grid"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
    </NamedIndividual>

    <!-- Individuals with 3 structures -->
    <NamedIndividual rdf:about="http://example.org/graph_301">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/grid"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
        <ns1:has_structure rdf:resource="http://example.org/wheel"/>
    </NamedIndividual>
    <NamedIndividual rdf:about="http://example.org/graph_302">
        <rdf:type rdf:resource="http://example.org/Graph"/>
        <ns1:has_structure rdf:resource="http://example.org/wheel"/>
        <ns1:has_structure rdf:resource="http://example.org/grid"/>
        <ns1:has_structure rdf:resource="http://example.org/house"/>
    </NamedIndividual>
</rdf:RDF>


```

**Comment by Demirrr** [comment]
Unsure what is the problem here.

> The model is supposed to generate a Description Logic expression that describes the Graph class having more than 2 has_structure relationships with instances of SuperStructure

Why do you think that the model is supposed to ...?

**Comment by AjayUmakanth** [comment]
In this scenario, each individual Graph is named in the format graph_n01, graph_n02, etc., where the n indicates the number of has_structure relationships connected to that graph. Specifically, the number of structures connected to each Graph individual is represented by the suffix of the name (e.g., graph_001 has 0 structures, graph_101 has 1 structure, and so on).

In the code, I am considering positive examples as Graph individuals with more than 2 structures connected to them, while all other Graphs with 0, 1, or 2 structures are considered negative examples.

To explain this classification in Description Logic (DL), we need a concept expression that captures the graphs with more than 2 structures, as the one suggested by me. As a result, the accuracy and F1 score would both be 1.0

**Comment by Demirrr** [comment]
I don't think that I understand your application or your reasoning. Importantly, your expected logic is not a valid DL expression, is it ?

Regardless, you may like to use other learners

**Comment by heindorf** [comment]
It seems the class expression `Graph ⊓ (≥ 3 has_structure.SuperStructure)` is evaluated incorrectly. The function `individuals_set` from `knowledge_base.py` returns the empty set yielding the metrics {'Accuracy': 0, 'Recall': 0, 'Precision': 0, 'F1': 0}. This not only affects EvoLearner, but CELOE, too (and probably other class expression learners).

@AjayUmakanth will provide additional details shortly.

**Comment by AjayUmakanth** [comment]
I think that there is an issue in the evaluation of the concept. I tried debugging this and I think there is an issue in the 'evaluate_concept' function. In that , 'individuals_set' is being called which gives an empty list for minimum cardinality restrictions. Due to that, all metrics,  'Accuracy', 'Recall', 'Precision' and 'F1' is 0.
I have created a class expression of the expected result

`from ontolearn.knowledge_base import KnowledgeBase
from ontolearn.learning_problem import PosNegLPStandard
from owlapy.owl_property import OWLObjectProperty
from owlapy.class_expression import OWLClass, OWLObjectMinCardinality, OWLObjectIntersectionOf
from owlapy.iri import IRI
from owlapy.render import DLSyntaxObjectRenderer
from ontolearn.metrics import F1, Accuracy, Precision, Recall

# Define paths and namespace
owl_file_path = "./customDblp.owl" #Replace with the graph path in your local
namespace = "http://example.org/"

# Load the knowledge base
kb = KnowledgeBase(path=owl_file_path)

has_structure_property = OWLObjectProperty(IRI.create(f"{namespace}has_structure"))

individuals = list(kb.individuals())
positive_examples = []
negative_examples = []

# Classify based on number of structures
for individual in individuals:
    structure_count = sum(
        1 for _ in kb.get_object_property_values(individual, has_structure_property)
    )
    if structure_count >= 3:
        positive_examples.append(individual)
    else:
        negative_examples.append(individual)

# Convert to sets for learning problem
typed_pos = set(positive_examples)
typed_neg = set(negative_examples)

# Create learning problem
lp = PosNegLPStandard(pos=typed_pos, neg=typed_neg)

# Build hypothesis4
graph_class = OWLClass(IRI.create(f"{namespace}Graph"))
super_structure_class = OWLClass(IRI.create(f"{namespace}SuperStructure"))

min_three_structures = OWLObjectMinCardinality(
    cardinality=3,
    property=has_structure_property,
    filler=super_structure_class
)

hypothesis = OWLObjectIntersectionOf([
    graph_class,
    min_three_structures
])


print("Positives:")
for example in positive_examples:
    print(example)

print("Negatives:")
for example in negative_examples:
    print(example)
    
# Print the hypothesis in Description Logic syntax
renderer = DLSyntaxObjectRenderer()
print("Hypothesis:")
print(renderer.render(hypothesis))

# Evaluate the hypothesis
metrics = [Accuracy(), Recall(), Precision(), F1()]
evaluation = {}
for metric in metrics:
    evaluated_concept = kb.evaluate_concept(hypothesis, metric, lp)
    evaluation[metric.name] = evaluated_concept.q

print("\nEvaluation Results:")
for metric_name, value in evaluation.items():
    print(f"{metric_name}: {value}")

# Get predictions for all individuals
predictions = model.predict(
    individuals=list(typed_pos | typed_neg),
    hypotheses=hypothesis
)
print("\nPredictions:")
print(predictions)`

This is the output
`
Positives:
OWLNamedIndividual(IRI('http://example.org/', 'graph_302'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_301'))
Negatives:
OWLNamedIndividual(IRI('http://example.org/', 'graph_204'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_103'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_201'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_102'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_202'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_001'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_002'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_203'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_101'))
OWLNamedIndividual(IRI('http://example.org/', 'graph_200'))
Hypothesis:
Graph ⊓ (≥ 3 has_structure.SuperStructure)

Evaluation Results:
Accuracy: 0
Recall: 0
Precision: 0
F1: 0

Predictions:
           Graph ⊓ (≥ 3 has_structure.SuperStructure)
graph_204                                         0.0
graph_103                                         0.0
graph_201                                         0.0
graph_202                                         0.0
graph_102                                         0.0
graph_001                                         0.0
graph_301                                         0.0
graph_002                                         0.0
graph_203                                         0.0
graph_302                                         0.0
graph_101                                         0.0
graph_200                                         0.0
`



Also, I also tried to execute the code using the latest example you provided. You can see that for the minimum cardinality it is always providing 0. BTW, I am using the same owlapy version that you are using

`from owlapy.owl_ontology_manager import SyncOntologyManager, OntologyManager
from owlapy.owl_reasoner import StructuralReasoner, SyncReasoner
from owlapy.owl_property import OWLObjectProperty
from owlapy.class_expression import OWLClass, OWLObjectMinCardinality, OWLObjectIntersectionOf, OWLObjectMaxCardinality
from owlapy.render import owl_expression_to_dl

cwr_reasoner=StructuralReasoner(ontology=OntologyManager().load_ontology(path="./customDblp.owl"))

A=OWLClass('http://example.org/Graph')
B=OWLObjectMinCardinality(cardinality=3, property=OWLObjectProperty('http://example.org/has_structure'), filler=OWLClass('http://example.org/SuperStructure'))
C=OWLObjectMaxCardinality(cardinality=3, property=OWLObjectProperty('http://example.org/has_structure'), filler=OWLClass('http://example.org/SuperStructure'))
AB = OWLObjectIntersectionOf([A, B])
AC = OWLObjectIntersectionOf([A, C])

print(f"{owl_expression_to_dl(A)}\t{len(set(cwr_reasoner.instances(A)))}")
print(f"{owl_expression_to_dl(B)}\t{len(set(cwr_reasoner.instances(B)))}")
print(f"{owl_expression_to_dl(C)}\t{len(set(cwr_reasoner.instances(C)))}")
print(f"{owl_expression_to_dl(AB)}\t{len(set(cwr_reasoner.instances(AB)))}")
print(f"{owl_expression_to_dl(AC)}\t{len(set(cwr_reasoner.instances(AC)))}")
`
Ouptut
`Graph	12
≥ 3 has_structure.SuperStructure	0
≤ 3 has_structure.SuperStructure	12
Graph ⊓ (≥ 3 has_structure.SuperStructure)	0
Graph ⊓ (≤ 3 has_structure.SuperStructure)	12`

**Comment by Demirrr** [comment]
Since the issue doen't related to a particular learning algorithm, I will be closing this issue. 

You can very well use protege to doublecheck the validitiy of the expected retreived instances
